### PR TITLE
Inference: reconstruct prediction as 3D nifti file

### PR DESCRIPTION
**Pull request by charleygros**
#### 0. Reconstruct 3D volume from slice-wise prediction
- Implemented in `ivadomed/utils`, `save_nii()`, called in `ivadomed/main`
- use `nibabel` library to re-orient the data accordingly to the original GT data
- nii files saved in `log_folder/pred_maks/`.

#### 1. Undo transformations after prediction
- `ToTensor`
- `NormaliseInstance`
- `ROICrop2D`
- `CenterCrop2D`
- `Resample`

#### 2. New testing functions
- `testing/test_inference`
- `testing/test_dilateGT`
- `testing/test_undoTrans`

#### 3.Unet model in testing data
- `testing_data/model_unet_test.pt`

### CODE REVIEW

File: ivadomed/transforms.py [hunk 1]
```diff
@@ -218,7 +290,10 @@ def __call__(self, sample):
 
             # post-processing
             gt_pp = self.post_processing(gt_data_np, gt_holes, gt_holes_bin, gt_dil)
```
[the remark this answers is not in the document]
Comment (AnBucquet) [reply ivadomed/transforms.py]: @charleygros it depends on which transformations you want to use after. 
For example, both [`ElasticTransform`](https://github.com/neuropoly/medicaltorch/blob/master/medicaltorch/transforms.py#L653-L662) and [`RandomAffine`](https://github.com/neuropoly/medicaltorch/blob/master/medicaltorch/transforms.py#L529-L536) will change de GT type. If that's a problem, we can change it?

File: ivadomed/utils.py [hunk 2]
```diff
@@ -33,6 +34,38 @@ def get_results(self):
         return res_dict
 
 
+def save_nii(data_lst, z_lst, fname_ref, fname_out, slice_axis):
+    nib_ref = nib.load(fname_ref)
```
Comment (jcohenadad) [inline ivadomed/utils.py]: it would be useful to add a docstrings as this function seems to do more than just saving a nifti file (otherwise calling nibabel would suffice)

File: testing/test_dilateGT.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+cudnn.benchmark = True
+
+GPU_NUMBER = 5
+BATCH_SIZE = 8
+DROPOUT = 0.4
+BN = 0.1
+N_EPOCHS = 10
+INIT_LR = 0.01
+PATH_BIDS = 'testing_data/'
```
Comment (jcohenadad) [inline testing/test_dilateGT.py]: throughout the code i would remove "/" to anticipate Windows compatibility and use `os.path.join()` that can deal with OS-specific separators

File: ivadomed/transforms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (charleygros) [reply ivadomed/transforms.py]: Right, all good then. Thanks

**Comment by charleygros** [comment]
Had kind of hard time jungling with the different conventions between PIL, numpy and pytorch... ... I do think that the `undo_transformations` are now correct.

**Comment by charleygros** [comment]
Testing of undo-transformations: comparing the original data and the data transformed-untransformed:
```
>>> from testing.test_undoTrans import test_undo
>>> test_undo(contrast='T2star', tol=3)
using GPU number 5
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1521.33it/s]

CenterCrop2D_ToTensor_NoramlizeInstance
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1640.64it/s]
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.

ROICrop2D_ToTensor_NoramlizeInstance
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1637.12it/s]
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.

Resample_CenterCrop2D_ToTensor_NoramlizeInstance
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1994.91it/s]
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.

Resample_ROICrop2D_ToTensor_NoramlizeInstance
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1970.08it/s]
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data content (tol: 3 vox.): checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
	Data shape: checked.
```

File: ivadomed/main.py [hunk 4]
```diff
@@ -615,7 +618,7 @@ def cmd_test(context):
 
     print(f"Loaded {len(ds_test)} {context['slice_axis']} slices for the test set.")
     test_loader = DataLoader(ds_test, batch_size=context["batch_size"],
-                             shuffle=True, pin_memory=True,
+                             shuffle=False, pin_memory=True,
```
Comment (charleygros) [inline ivadomed/main.py]: `shuffle=False` --> in order to perform the prediction on consecutive slices, easier to further reconstruct the 3D vol.

File: ivadomed/transforms.py [hunk 5]
```diff
@@ -24,14 +34,30 @@ def __init__(self, wspace, hspace,
         super().__init__(wspace, hspace, interpolation, labeled)
 
     def resample_bin(self, data, wshape, hshape, thr=0.5):
-        data = data.resize((wshape, hshape), resample=self.interpolation)
-        np_data = np.array(data)
-        np_data[np_data > thr] = 255
-        np_data[np_data <= thr] = 0
-        np_data = np_data.astype(np.uint8)
-        data = Image.fromarray(np_data, mode='L')
+        data = data.resize((wshape, hshape), resample=Image.NEAREST)
```
Comment (charleygros) [inline ivadomed/transforms.py]: More robust for binary data. Tested on 10 subjects.

File: ivadomed/main.py [hunk 6]
```diff
@@ -656,6 +665,35 @@ def cmd_test(context):
             else:
                 preds = model(test_input)
 
+        # WARNING: sample['gt'] is actually the pred in the return sample
+        # implementation justification: the other option: rdict['pred'] = preds would require to largely modify mt_transforms
+        rdict = {}
+        rdict['gt'] = preds.cpu()
```
Comment (charleygros) [inline ivadomed/main.py]: Here I overwrite the GT by the preds in order to benefit from `gt_metadata` & cie during `save_nii`.

File: ivadomed/transforms.py [hunk 7]
```diff
@@ -12,7 +12,17 @@
...
-    return ['DilateGT', 'ROICrop2D', 'Resample']
+    return ['DilateGT', 'ROICrop2D', 'Resample', 'NormalizeInstance', 'ToTensor', 'CenterCrop2D']
+
+
+class UndoCompose(object):
+    def __init__(self, compose):
+        self.transforms = compose.transforms
+
+    def __call__(self, img):
+        for t in self.transforms[::-1]:
```
Comment (charleygros) [inline ivadomed/transforms.py]: List of transforms reversed.

File: ivadomed/transforms.py [hunk 8]
```diff
@@ -44,8 +70,8 @@ def __call__(self, sample):
         hfactor = hzoom / self.hspace
         wfactor = wzoom / self.wspace
 
-        hshape_new = int(hshape * hfactor)
-        wshape_new = int(wshape * wfactor)
+        hshape_new = int(round(hshape * hfactor))
```
Comment (charleygros) [inline ivadomed/transforms.py]: Better reproducibility.

File: ivadomed/transforms.py [hunk 9]
```diff
@@ -63,7 +89,40 @@ def __call__(self, sample):
         return sample
 
 
-class ROICrop2D(mt_transforms.CenterCrop2D):
+class NormalizeInstance(mt_transforms.NormalizeInstance):
+    """This class extends mt_transforms.Normalize"""
+    def undo_transform(self, sample):
+        return sample
```
Comment (charleygros) [inline ivadomed/transforms.py]: Here we skip the undo for the images as we no longer care about the intensities at this stage.

File: ivadomed/transforms.py [hunk 10]
```diff
@@ -211,6 +286,8 @@ def post_processing(self, arr_in, arr_soft, arr_bin, arr_dil):
     def __call__(self, sample):
         gt_data = sample['gt']
         gt_data_np = np.array(gt_data)
+        # binarize for processing
+        gt_data_np = (gt_data_np > 0.5).astype(np.int_)
```
Comment (charleygros) [inline ivadomed/transforms.py]: Reg. recent changes with PIL

File: ivadomed/transforms.py [hunk 11]
```diff
@@ -221,7 +298,10 @@ def __call__(self, sample):
 
             # post-processing
             gt_pp = self.post_processing(gt_data_np, gt_holes, gt_holes_bin, gt_dil)
-            gt_out = gt_data_np.astype(np.float64)
+
+            # mask with ROI
+            if sample['roi'] is not None:
+                gt_pp[np.array(sample['roi']) == 0] = 0.0
```
Comment (charleygros) [inline ivadomed/transforms.py]: post-processing

File: testing/test_inference.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                                  slice_axis=SLICE_AXIS,
+                                  transform=val_transform,
+                                  multichannel=False,
+                                  slice_filter_fn=SliceFilter(filter_empty_input=True,
+                                                                filter_empty_mask=False))
+
+    ds_test.filter_roi(nb_nonzero_thr=10)
+
+    if film_bool:  # normalize metadata before sending to network
+        print('FiLM inference not implemented yet.')
```
Comment (charleygros) [inline testing/test_inference.py]: todo at some point

File: testing/test_inference.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+            for k in batch.keys():
+                rdict[k] = batch[k][smp_idx]
+            rdict_undo = val_undo_transform(rdict)
+
+            fname_ref = rdict_undo['input_metadata']['gt_filename']
+            if pred_tmp_lst and (fname_ref != fname_tmp or (i == len(test_loader)-1 and smp_idx == len(batch['gt'])-1)):  # new processed file
+                # save the completely processed file as a nii
+                fname_pred = os.path.join(PATH_OUT, fname_tmp.split('/')[-1])
+                fname_pred = fname_pred.split('manual.nii.gz')[0] + 'pred.nii.gz'
+                save_nii(pred_tmp_lst, z_tmp_lst, fname_tmp, fname_pred, SLICE_AXIS)
```
Comment (AnBucquet) [inline testing/test_inference.py]: @charleygros I've changed `model_unet_test.pt` and fixed the problem with Cuda in `test_inference.py`, but I got a new error with [save_nii(pred_tmp_lst, z_tmp_lst, fname_tmp, fname_pred, SLICE_AXIS)](https://github.com/neuropoly/ivado-medical-imaging/blob/cg/reconstruct-3D/testing/test_inference.py#L150).
 
--> here: [pytest](https://github.com/neuropoly/ivado-medical-imaging/runs/346353620#step:6:156).

**Comment by andreanne-lemay** [comment]
Currently this PR is not compatible with multichannel. I can do the changes to make it compatible (minor changes: use only one channel to output the images), but I would wait for @AnBucquet 's PR on medicaltorch to be merged to avoid making the work twice since Anthime's PR will change the tensor form (list vs a stacked tensor). Meanwhile, a condition `if multichannel` could be added to avoid an error, so at least the command `test` can be used in multichannel.

File: ivadomed/main.py [hunk 14]
```diff
@@ -656,6 +665,35 @@ def cmd_test(context):
...
+                # save the completely processed file as a nii
+                fname_pred = os.path.join(path_3Dpred, fname_tmp.split('/')[-1])
+                fname_pred = fname_pred.split('manual.nii.gz')[0] + 'pred.nii.gz'
+                save_nii(pred_tmp_lst, z_tmp_lst, fname_tmp, fname_pred, axis_dct[context['slice_axis']])
+                # re-init pred_stack_lst
+                pred_tmp_lst, z_tmp_lst = [], []
+                # TODO compute image-based metrics
+
+            # add new sample to pred_tmp_lst
+            pred_tmp_lst.append(np.array(rdict_undo['input']))
```
Comment (andreanne-lemay) [inline ivadomed/main.py]: pred_tmp_lst should contain the prediction, should it be rdict_undo['gt'] instead of rdict_undo['input']?

Comment (charleygros) [reply ivadomed/main.py]: ahah yes absolutely! I used input for my testing, but yes of course gt :)

File: ivadomed/main.py [hunk 15]
```diff
@@ -671,12 +671,17 @@ def cmd_test(context):
         rdict['gt'] = preds.cpu()
         batch.update(rdict)
 
+        if batch["input"].shape[0] > 1:
```
Comment (charleygros) [inline ivadomed/main.py]: It seems that we are entering in this if condition when we are in monochannel --> will look at it

Comment (andreanne-lemay) [reply ivadomed/main.py]: oups yes clearly because of the batch size should be shape[1]
